crates/circuit/src/packed_instruction.rs

```diff
@@ -39,132 +42,375 @@ enum PackedOperationType {
 ///                           Standard gate, stored inline as a u8. --+     +-- Discriminant.
 ///
-/// Python object:
-/// 0b_PPPPPPPP_PPPPPPPP_PPPPPPPP_PPPPPPPP_PPPPPPPP_PPPPPPPP_PPPPPPPP_PPPPP10
-///    |------------------------------------------------------------------|||
+/// StandardInstructionBits:
+/// 0b_DDDDDDDD_DDDDDDDD_DDDDDDDD_DDDDDDDD_xxxxxxxx_xxxxxxxx_xxxxxSSS_SSSSS001
+///    |---------------------------------|                        |-------||-|
+///                    |                                              |     |
+///                    +-- An optional 32 bit immediate value.        |     |
+///               Standard instruction type, stored inline as a u8. --+     +-- Discriminant.
+///
+///     Optional immediate value:
+///     Depending on the variant of the standard instruction type, a 32 bit
+///     inline value may be present. Currently, this is used to store the
+///     number of qubits in a Barrier and the unit of a Delay.
+///
+/// PointerBits (64-bit systems):
+/// 0b_PPPPPPPP_PPPPPPPP_PPPPPPPP_PPPPPPPP_PPPPPPPP_PPPPPPPP_PPPPPPPP_PPPP011
+///    |-----------------------------------------------------------------||-|
 ///                                   |                                    |
 ///    The high 62 bits of the pointer.  Because of alignment, the low 3   |   Discriminant of the
-///    bits of the full 64 bits are guaranteed to be zero (so one marked   +-- enumeration.  This
-///    `P` is always zero here), so we can retrieve the "full" pointer by      is 0b10, which means
-///    taking the whole `usize` and zeroing the low 3 bits, letting us         that this points to
-///    store the discriminant in there at other times.                         a `PyInstruction`.
-/// ```
+///    bits of the full 64 bits are guaranteed to be zero so we can        +-- enumeration.  This
+///    retrieve the "full" pointer by taking the whole `u64` and zeroing       is 0b011, which means
+///    the low 3 bits, letting us store the discriminant in there at other     that this points to
+///    times.                                                                  a `PyInstruction`.
 ///
-/// There is currently one spare bit that could be used for additional metadata, if required.
+/// PointerBits (32-bit systems)
+/// 0b_PPPPPPPP_PPPPPPPP_PPPPPPPP_PPPPPPPP_xxxxxxxx_xxxxxxxx_xxxxxxxx_xxxx011
+///    |---------------------------------|                                |-|
+///                    |                                                   |
+///    On 32-bit systems, the upper 32-bits are used to store the full     +-- Discriminant. This is
+///    address. This gives the compiler the best chance at eliminating         0b100, so in this
+///    the otherwise required bit-shift needed to read the pointer.            case, this points to
+///                                                                            a `PyOperation`.
```

Comment:
If anything, 32-bit CPUs with memory indirect addressing modes would be able to more efficiently dereference the pointer if it doesn't need shifting and masking, right?

And, we can't use a 3-bit mask on 32-bit systems since the addresses would be 4-byte aligned.

The idea here was that we're making life worse for 32-bit systems by making `PackedOperation` always 64 bits. I wanted to therefore do as much as easily possible to curb that. I don't think the separate layout is a big deal in terms of complexity, and it doesn't need any special consideration externally to `PointerBits`.

Reply:
If the dtructs being pointed to aren't already forcibly 8-byte aligned, you can stick an `align(8)` on them - I thought I did it already, but if I didn't, we'd always planned to.

Yeah, I'm not arguing that this wouldn't be ever so slightly more efficient for 32-bit. I'm just saying that it comes with a lot of complexity to achieve that, and I don't think it's justified. All the pointer forms are the Python-interaction ones already, which means the performance is already going to be naff.